80 dataset drops (also 82, 83, 90)

philswatton opened the pull request:
This PR adds several functions, mainly in src/locomoset/data/preprocess, to:

- Resolves #80 by adding two functions, `drop_images` and `drop_images_by_label`
  - `drop_images` takes `keep_size` as an argument, which can be either an int or a float between 0 and 1 specifying how many images to keep. The drops are stratified by label so pre-existing label balance will be preserved
  - `drop_images_by_label` takes `keep_labels` as an argument, which is a list containing all labels to keep. All images without these labels are dropped.
  - `apply_dataset_mutations` takes both arguments and applies both functions (dropping by label first, randomly dropping images second)
  - Under the hood, these functions use `_mutate_dataset` which takes a dataset or dataset dict, a function, and kwargs to that function, then applies it. This makes it easier to ignore what kind of object is being worked with
- Resolves #82 by refactoring the experiment to use a config class, rather than a dict
- Resolves #83 by adding train/val/test splits via `create_data_splits`. This will automatically detect whether a single dataset is being split into three, a dict of two datasets needs val or test to be added (i.e. train must always exist - the val or test set will be created from train), or if it is already split three ways, and return a dataset dict with all three splits
  - Note this is a change in behaviour: `train_split` and `val_split` (and by extension, the new `test_split` should NOT contain the same names. If these splits exist, they should match the names in the dataset: otherwise they are used to create new splits with the corresponding names
- Resolves #90 by ensuring top-level configs are consistent and use `wandb_args`, not `wandb`
- Adds tests for all new functions

This involved some refactoring in places, as running training and metrics experiments via configs needed to be rewritten. Given the same arguments in a config (in terms of dataset name, split names, val/test sizes), both training and metrics will use

Unit tests have been updated in some places where relevant

Not implemented in this PR:
- #81: the configs are a *little* less permissive where I've been changing stuff, but I've left actually doing this for a PR in the new year
- actually using the test set

philswatton committed:
image and label dropping

philswatton committed:
basic idea for usage

philswatton committed:
renamed fn

philswatton committed:
adding fns to experiment

philswatton committed:
resolved merge conflict

philswatton committed:
Merge branch 'develop' into 80-dataset-drops

philswatton committed:
handling of labels - need to check if this breaks training

philswatton committed:
removed random [train] subset

philswatton committed:
data mmutation fns + train/val/test added

philswatton committed:
fns added to fns that take configs

philswatton committed:
experiment now uses config

philswatton committed:
new dataset args added for training

philswatton committed:
new args added to configs

philswatton committed:
wandb -> wandb_args

philswatton committed:
only 3 tests failing now

philswatton committed:
all tests passing

philswatton committed:
unit tests added for new functionality

philswatton committed:
metric_kwargs enabled at config level

philswatton committed:
docstrings

philswatton committed:
README updated

philswatton committed:
fixing docstrings, args, tests

jack89roberts committed:
ensure self.dataset is set

jack89roberts committed:
Merge branch 'develop' into 80-dataset-drops

jack89roberts committed:
include test metrics in training

jack89roberts committed:
fix tests

jack89roberts committed:
rename a few things

jack89roberts commented:
Keeping track of a few things here:

- I refactored the splitting a bit, mostly so that we keep a large test dataset in all cases (i.e. don't scale it down along with the train/val dataset sizes). The size of the `train` split is now controlled by `n_samples` (as used by the metrics). The `val` split is further subsetted before training starts to make it at most one-quarter of the training dataset size - best to keep it small since it's only for debugging whether the training has gone ok now.

- I've made it so the training jobs log the test dataset accuracy (once at the end).

- The LoCoMoSeT metrics will be computed on the train dataset only now, rather than train + val. I felt it was weird to potentially be computing the metrics on more (or different) images than the model sees in training. Could be convinced to revert that.

- **_[Should only an issue for sub-setting existing datasets, should not be an issue for the binary datasets created elsewhere]_** After dropping labels the `ClassLabel` in the dataset (`dataset["train"].features["label"]`) still has all the original labels and the class IDs correspond to the original dataset (so e.g. they may be 7 and 23 instead of 0 and 1). This means we will end up training a model with e.g. a 100 class classification head but it will only see 2 of those classes during training (and the metrics will only be computed on those classes). Subsetting a `ClassLabel` instance and updating the dataset labels to reflect isn't a built-in feature of HF datasets as far as I could tell.

- `keep_labels` is not nested under `dataset_args` in the top level configs but is in the individual configs (not attempted to fix yet, partially since `keep_labels` is set to be one of the things that can be looped over and unsure how messy that would get).

jack89roberts committed:
refactor splits so test set size not impacted by no. samples

- train and metrics now both use `n_samples` - training to define size of train split, metrics as before

- keep_size removed as a config argument in favour of `n_samples`

- test set created from whole dataset (or original test split if available) and is not subsetted further if the train/val splits are reduced in size

- val split size is set to whichever is smaller between the original val split size and the training dataset (e.g. to avoid training on 500 samples but validating on 50000)

jack89roberts committed:
compute metrics on train split only

To discuss as I realise this was what we planned, but it felt strange to be having metrics based on images that won't be used to train the model.

jack89roberts committed:
further reduce size of val set

The val set is now for debugging purposes only, to ensure nothing has gone very wrong in training.

This should really be a parameter added back into the config (bit like the keep size just removed...)

jack89roberts committed:
bugfix configs

jack89roberts committed:
remove some old args

eddableheath committed:
updated gitignore

jack89roberts committed:
check for ClassLabel instance rather than string

jack89roberts committed:
use same default batch size for get_features

jack89roberts commented:
Going to merge this so we don't have a messy mix of PRs but can talk through the above when you're back @philswatton
